client/state/sites/actions.js

```diff
@@ -115,55 +115,52 @@ export function requestSites() {
  * a site.
  *
  * @param {number|string} siteFragment Site ID or slug
- * @param {boolean} forceWpcom explicitly get info from WPCOM vs Jetpack site
  * @returns {Function}              Action thunk
  */
-export function requestSite( siteFragment, forceWpcom = false ) {
+export function requestSite( siteFragment ) {
+	function doRequest( forceWpcom ) {
+		const query = { apiVersion: '1.2' };
+		if ( forceWpcom ) {
+			query.force = 'wpcom';
+		}
+
+		const siteFilter = config( 'site_filter' );
+		if ( siteFilter.length > 0 ) {
+			query.filters = siteFilter.join( ',' );
+		}
+
+		return wpcom.site( siteFragment ).get( query );
+	}
+
 	return ( dispatch ) => {
-		dispatch( {
-			type: SITE_REQUEST,
-			siteId: siteFragment,
+		dispatch( { type: SITE_REQUEST, siteId: siteFragment } );
+
+		const result = doRequest( false ).catch( ( error ) => {
+			// if there is Jetpack JSON API module error, retry with force: 'wpcom'
+			if (
+				error?.status === 403 &&
+				error?.message === 'API calls to this blog have been disabled.'
+			) {
+				return doRequest( true );
+			}
+
+			return Promise.reject( error );
 		} );
-		const siteFilter = config( 'site_filter' );
-		return wpcom
-			.site( siteFragment )
-			.get( {
-				apiVersion: '1.2',
-				filters: siteFilter.length > 0 ? siteFilter.join( ',' ) : undefined,
-				force: forceWpcom ? 'wpcom' : undefined,
-			} )
+
+		result
 			.then( ( site ) => {
 				// If we can't manage the site, don't add it to state.
-				if ( ! ( site && site.capabilities ) ) {
-					return dispatch( {
-						type: SITE_REQUEST_FAILURE,
```

Comment:
I wonder if any of the consumer code was relying on the fact that we were dispatching a failure in the case when the user doesn't have enough permissions. I wasn't able to verify that at a single glance. Have you checked it yourself somehow?

Reply:
Yes, I verified that -- the `SITE_REQUEST_FAILURE` and `SITE_REQUEST_SUCCESS` action only reset a `requesting` flag to `false`, they have no other function. They are identical to each other.

There is a special case in Jetpack Connect Authorize where the `FAILURE` action is used to set a `clientNotResponding` flag. But that component issues its own `wpcom.site().get()` request (in the `isUserConnected` action) and also dispatches its own `SITE_REQUEST_FAILURE` action. Changes in `state/sites` shouldn't affect that.